src/passes/Print.cpp

```diff
@@ -2665,36 +2665,62 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
       printTableHeader(curr);
       o << maybeNewLine;
     }
-    for (auto& segment : curr->segments) {
-      // Don't print empty segments
-      if (segment.data.empty()) {
-        continue;
-      }
-      doIndent(o, indent);
-      o << '(';
-      printMedium(o, "elem ");
 
+    ModuleUtils::iterTableSegments(
+      *currModule, curr->name, [&](ElementSegment* segment) {
+        printElementSegment(segment);
+      });
+  }
+  void visitElementSegment(ElementSegment* curr) {
+    if (curr->table.is()) {
+      return;
+    }
+
+    printElementSegment(curr);
+  }
+  void printElementSegment(ElementSegment* curr, bool isDeclarative = false) {
+    // Don't print empty segments
+    if (curr->data.empty()) {
+      return;
+    }
+    doIndent(o, indent);
+    o << '(';
+    printMedium(o, "elem");
+    if (curr->hasExplicitName) {
+      o << ' ';
+      printName(curr->name, o);
+    }
+
+    if (curr->table.is()) {
       // TODO(reference-types): check for old-style based on the complete spec
       if (currModule->tables.size() > 1) {
         // tableuse
-        o << '(';
-        printMedium(o, "table ");
-        printName(curr->name, o);
-        o << ") ";
+        o << " (table ";
+        printName(curr->table, o);
+        o << ")";
       }
 
-      visit(segment.offset);
+      o << ' ';
+      visit(curr->offset);
 
       if (currModule->tables.size() > 1) {
-        o << " func";
+        o << ' ';
+        TypeNamePrinter(o, currModule).print(HeapType::func);
       }
-
-      for (auto name : segment.data) {
+    } else {
+      if (isDeclarative) {
```

Comment:
Well that would involve making an `ElementSegment` for the declared functions, which I suspect isn't worth it, although it will make the code simpler.

Reply:
Ok, assuming that we don't want to make an artificial `ElementSegment` just for printing the declarative segments, it looks like `isDeclarative` should not be necessary in this function.

Reply:
Right, it's now removed.